File: ChatGPT/Direct_Attempt/clojure/output/output_15b.clj
;; ChatGPT-generated solution will go here.
(ns output_15b
  (:require [clojure.string :as str]
            [clojure.set :as set]))

(defn widen-tile [ch]
  (case ch
    \# [\# \#]
    \O [\] \[] ;; represent wide box as closing and opening for visual order
    \. [\. \.]
    \@ [\@ \.]
    ;; fallback
    [ch ch]))

(defn widen-map [lines]
  (vec (mapcat (fn [line]
                 [(vec (mapcat widen-tile line))])
               lines)))

(defn parse-moves [lines]
  (->> lines
       (drop-while #(not (str/starts-with? % "<")))
       (apply str)
       (remove #{\newline})
       (map {\^ [-1 0] \v [1 0] \< [0 -1] \> [0 1]})))

(defn initial-state [grid]
  (let [rows (count grid)
        cols (count (first grid))]
    (loop [r 0, c 0]
      (cond
        (= r rows) nil
        (= c cols) (recur (inc r) 0)
        (= (get-in grid [r c]) \@)
        {:robot [r c]
         :boxes (set (for [i (range rows)
                           j (range 0 cols 2)
                           :when (and (= (get-in grid [i j]) \[)
                                      (= (get-in grid [i (inc j)]) \]))]
                      [i j]))
         :walls (set (for [i (range rows)
                           j (range cols)
                           :when (= (get-in grid [i j]) \#)]
                      [i j]))
         :rows rows
         :cols cols}
        :else (recur r (inc c))))))

(defn box-occupied? [boxes [r c]]
  (or (boxes [r c]) (boxes [r (dec c)]) (boxes [r (inc c)])))

(defn can-move-box? [state [r c] [dr dc]]
  (let [dst [(+ r dr) (+ c dc)]
        dst2 [(+ r dr) (+ c dc 1)]]
    (and (not (state :walls dst))
         (not (state :walls dst2))
         (not (box-occupied? (state :boxes) dst))
         (not (box-occupied? (state :boxes) dst2)))))

(defn move-box [boxes src dir]
  (let [[dr dc] dir
        dst [(+ (first src) dr) (+ (second src) dc)]]
    (-> boxes
        (disj src)
        (conj dst))))

(defn move-robot [state dir]
  (let [robot (:robot state)
        [dr dc] dir
        dest [(+ dr (first robot)) (+ dc (second robot))]
        box-in-way? (box-occupied? (:boxes state) dest)
        new-boxes (if box-in-way?
                    (let [src (if ((:boxes state) dest)
                                dest
                                [(first dest) (dec (second dest))])]
                      (if (can-move-box? state src dir)
                        (move-box (:boxes state) src dir)
                        nil))
                    (:boxes state))]
    (if (or (state :walls dest)
            (and box-in-way? (nil? new-boxes)))
      state
      (assoc state :robot dest :boxes new-boxes))))

(defn simulate [grid moves]
  (loop [state (initial-state grid)
         mv moves]
    (if (empty? mv)
      state
      (recur (move-robot state (first mv)) (rest mv)))))

(defn gps [[r c]]
  (+ (* 100 r) c))

(defn part [lines]
  (let [grid (widen-map lines)
        moves (parse-moves lines)
        final-state (simulate grid moves)]
    (reduce + (map gps (:boxes final-state)))))

(defn -main [& args]
  (let [filename (first args)
        lines (str/split-lines (slurp filename))]
    (println (part lines))))

File: ChatGPT/Direct_Attempt/python/output/output_15b.py
# ChatGPT-generated solution will go here.
def main():
    with open("input.txt") as f:
        lines = [line.rstrip('\n') for line in f]

    raw_map = []
    moves = ""
    reading_map = True

    for line in lines:
        if reading_map:
            if line.strip() == "":
                reading_map = False
            else:
                raw_map.append(line)
        else:
            moves += line.strip()

    orig_rows = len(raw_map)
    orig_cols = len(raw_map[0])
    rows = orig_rows
    cols = orig_cols * 2

    grid = [[''] * cols for _ in range(rows)]
    for r in range(rows):
        c_out = 0
        for c in range(orig_cols):
            ch = raw_map[r][c]
            if ch == '#':
                grid[r][c_out] = '#'
                grid[r][c_out + 1] = '#'
            elif ch == '.':
                grid[r][c_out] = '.'
                grid[r][c_out + 1] = '.'
            elif ch == 'O':
                grid[r][c_out] = '['
                grid[r][c_out + 1] = ']'
            elif ch == '@':
                grid[r][c_out] = '@'
                grid[r][c_out + 1] = '.'
            c_out += 2

    # Locate robot
    for r in range(rows):
        for c in range(cols - 1):
            if grid[r][c] == '@':
                robot_r, robot_c = r, c
                grid[r][c] = '.'

    directions = {'^': (-1, 0), 'v': (1, 0), '<': (0, -1), '>': (0, 1)}

    for move in moves:
        dr, dc = directions[move]
        nr, nc = robot_r + dr, robot_c + dc

        if not (0 <= nr < rows and 0 <= nc < cols - 1):
            continue

        tgt1, tgt2 = grid[nr][nc], grid[nr][nc + 1]
        if tgt1 == '.' and tgt2 == '.':
            robot_r, robot_c = nr, nc
        elif tgt1 == '[' and tgt2 == ']':
            nnr, nnc = nr + dr, nc + dc
            if 0 <= nnr < rows and 0 <= nnc < cols - 1:
                after1, after2 = grid[nnr][nnc], grid[nnr][nnc + 1]
                if after1 == '.' and after2 == '.':
                    grid[nnr][nnc] = '['
                    grid[nnr][nnc + 1] = ']'
                    grid[nr][nc] = '.'
                    grid[nr][nc + 1] = '.'
                    robot_r, robot_c = nr, nc

    # Compute GPS sum
    total = 0
    for r in range(rows):
        for c in range(cols - 1):
            if grid[r][c] == '[' and grid[r][c + 1] == ']':
                total += r * 100 + c

    print(total)

if __name__ == "__main__":
    main()
